File: backend/services/lp_types.py
from __future__ import annotations

import re

LP_TYPE_INSTITUTIONAL = "유한책임조합원(기관투자자)"
LP_TYPE_INDIVIDUAL = "유한책임조합원(개인투자자)"
LP_TYPE_GP = "업무집행조합원(GP)"
LP_TYPE_SPECIAL_MOTAE = "특별조합원(모태)"
LP_TYPE_SPECIAL_NONG_MOTAE = "특별조합원(농모태)"
LP_TYPE_SPECIAL_GROWTH = "특별조합원(성장금융)"
LP_TYPE_SPECIAL_OTHER_POLICY = "특별조합원(기타정책자금)"

LP_TYPE_OPTIONS = (
    LP_TYPE_INSTITUTIONAL,
    LP_TYPE_INDIVIDUAL,
    LP_TYPE_GP,
    LP_TYPE_SPECIAL_MOTAE,
    LP_TYPE_SPECIAL_NONG_MOTAE,
    LP_TYPE_SPECIAL_GROWTH,
    LP_TYPE_SPECIAL_OTHER_POLICY,
)

LP_TYPE_GROUP_LIMITED = "유한책임조합원"
LP_TYPE_GROUP_GP = "업무집행조합원"
LP_TYPE_GROUP_SPECIAL = "특별조합원"

MIGRATION_LP_TYPE_OPTIONS = (
    LP_TYPE_INSTITUTIONAL,
    LP_TYPE_INDIVIDUAL,
    LP_TYPE_SPECIAL_MOTAE,
    LP_TYPE_SPECIAL_NONG_MOTAE,
    LP_TYPE_SPECIAL_GROWTH,
    LP_TYPE_SPECIAL_OTHER_POLICY,
)

_KEY_PATTERN = re.compile(r"[\s\-_()/]+")


def _normalize_key(value: str | None) -> str:
    text = (value or "").strip()
    if not text:
        return ""
    return _KEY_PATTERN.sub("", text).casefold()


def _build_alias_map() -> dict[str, str]:
    aliases: dict[str, str] = {}

    def register(canonical: str, *values: str) -> None:
        for value in (canonical, *values):
            key = _normalize_key(value)
            if key:
                aliases[key] = canonical

    register(
        LP_TYPE_INSTITUTIONAL,
        "institutional",
        "기관투자자",
        "기관",
        "법인",
        "corporate",
        "limitedpartnerinstitutional",
    )
    register(
        LP_TYPE_INDIVIDUAL,
        "individual",
        "개인투자자",
        "개인",
        "limitedpartnerindividual",
    )
    register(
        LP_TYPE_GP,
        "gp",
        "업무집행조합원",
        "공동업무집행",
        "공동gp",
        "cogp",
        "co-gp",
        "generalpartner",
    )
    register(
        LP_TYPE_SPECIAL_MOTAE,
        "모태",
        "motae",
        "motaefund",
    )
    register(
        LP_TYPE_SPECIAL_NONG_MOTAE,
        "농모태",
        "nongmotae",
        "nong-motae",
        "nongmotaefund",
    )
    register(
        LP_TYPE_SPECIAL_GROWTH,
        "성장금융",
        "growthfinance",
        "growth-finance",
    )
    register(
        LP_TYPE_SPECIAL_OTHER_POLICY,
        "특별조합원",
        "정책출자자",
        "정책자금",
        "기타정책자금",
        "정책",
        "정부기관",
        "government",
        "policy",
        "policymoney",
        "specialpartner",
    )
    return aliases


_LP_TYPE_ALIAS_MAP = _build_alias_map()
_LP_TYPE_OPTION_SET = set(LP_TYPE_OPTIONS)
_MIGRATION_LP_TYPE_SET = set(MIGRATION_LP_TYPE_OPTIONS)
_SPECIAL_LP_TYPE_SET = {
    LP_TYPE_SPECIAL_MOTAE,
    LP_TYPE_SPECIAL_NONG_MOTAE,
    LP_TYPE_SPECIAL_GROWTH,
    LP_TYPE_SPECIAL_OTHER_POLICY,
}


def normalize_lp_type(value: str | None) -> str | None:
    text = (value or "").strip()
    if not text:
        return None
    return _LP_TYPE_ALIAS_MAP.get(_normalize_key(text), text)


def coerce_lp_type(value: str | None) -> str:
    normalized = normalize_lp_type(value)
    if not normalized:
        raise ValueError("type must not be empty")
    if normalized not in _LP_TYPE_OPTION_SET:
        raise ValueError("unsupported LP type")
    return normalized


def normalize_lp_type_group(value: str | None) -> str | None:
    normalized = normalize_lp_type(value)
    if normalized in {LP_TYPE_INSTITUTIONAL, LP_TYPE_INDIVIDUAL}:
        return LP_TYPE_GROUP_LIMITED
    if normalized == LP_TYPE_GP:
        return LP_TYPE_GROUP_GP
    if normalized in _SPECIAL_LP_TYPE_SET:
        return LP_TYPE_GROUP_SPECIAL
    return normalized


def is_gp_lp_type(value: str | None) -> bool:
    return normalize_lp_type(value) == LP_TYPE_GP


def is_special_lp_type(value: str | None) -> bool:
    normalized = normalize_lp_type(value)
    return normalized in _SPECIAL_LP_TYPE_SET


def is_supported_lp_type(value: str | None) -> bool:
    normalized = normalize_lp_type(value)
    return normalized in _LP_TYPE_OPTION_SET


def is_supported_migration_lp_type(value: str | None) -> bool:
    normalized = normalize_lp_type(value)
    return normalized in _MIGRATION_LP_TYPE_SET

File: frontend/src/lib/lpTypes.ts
export const LP_TYPE_INSTITUTIONAL = '유한책임조합원(기관투자자)'
export const LP_TYPE_INDIVIDUAL = '유한책임조합원(개인투자자)'
export const LP_TYPE_GP = '업무집행조합원(GP)'
export const LP_TYPE_SPECIAL_MOTAE = '특별조합원(모태)'
export const LP_TYPE_SPECIAL_NONG_MOTAE = '특별조합원(농모태)'
export const LP_TYPE_SPECIAL_GROWTH = '특별조합원(성장금융)'
export const LP_TYPE_SPECIAL_OTHER_POLICY = '특별조합원(기타정책자금)'

export const LP_TYPE_OPTIONS = [
  LP_TYPE_INSTITUTIONAL,
  LP_TYPE_INDIVIDUAL,
  LP_TYPE_GP,
  LP_TYPE_SPECIAL_MOTAE,
  LP_TYPE_SPECIAL_NONG_MOTAE,
  LP_TYPE_SPECIAL_GROWTH,
  LP_TYPE_SPECIAL_OTHER_POLICY,
] as const

export type LpTypeValue = (typeof LP_TYPE_OPTIONS)[number]

export const DEFAULT_LP_TYPE: LpTypeValue = LP_TYPE_INSTITUTIONAL

export const LP_TYPE_SELECT_GROUPS: Array<{ label: string; options: ReadonlyArray<LpTypeValue> }> = [
  { label: '유한책임조합원', options: [LP_TYPE_INSTITUTIONAL, LP_TYPE_INDIVIDUAL] },
  { label: '업무집행조합원', options: [LP_TYPE_GP] },
  { label: '특별조합원', options: [LP_TYPE_SPECIAL_MOTAE, LP_TYPE_SPECIAL_NONG_MOTAE, LP_TYPE_SPECIAL_GROWTH, LP_TYPE_SPECIAL_OTHER_POLICY] },
]

const LP_TYPE_SET = new Set<string>(LP_TYPE_OPTIONS)

function normalizeKey(value: string | null | undefined): string {
  return (value || '').trim().toLowerCase().replace(/[\s\-_()/]+/g, '')
}

const LP_TYPE_ALIASES = new Map<string, LpTypeValue>([
  [normalizeKey(LP_TYPE_INSTITUTIONAL), LP_TYPE_INSTITUTIONAL],
  [normalizeKey('institutional'), LP_TYPE_INSTITUTIONAL],
  [normalizeKey('기관투자자'), LP_TYPE_INSTITUTIONAL],
  [normalizeKey('기관'), LP_TYPE_INSTITUTIONAL],
  [normalizeKey('법인'), LP_TYPE_INSTITUTIONAL],
  [normalizeKey('corporate'), LP_TYPE_INSTITUTIONAL],
  [normalizeKey(LP_TYPE_INDIVIDUAL), LP_TYPE_INDIVIDUAL],
  [normalizeKey('individual'), LP_TYPE_INDIVIDUAL],
  [normalizeKey('개인투자자'), LP_TYPE_INDIVIDUAL],
  [normalizeKey('개인'), LP_TYPE_INDIVIDUAL],
  [normalizeKey(LP_TYPE_GP), LP_TYPE_GP],
  [normalizeKey('gp'), LP_TYPE_GP],
  [normalizeKey('업무집행조합원'), LP_TYPE_GP],
  [normalizeKey('공동업무집행'), LP_TYPE_GP],
  [normalizeKey('co-gp'), LP_TYPE_GP],
  [normalizeKey('cogp'), LP_TYPE_GP],
  [normalizeKey(LP_TYPE_SPECIAL_MOTAE), LP_TYPE_SPECIAL_MOTAE],
  [normalizeKey('모태'), LP_TYPE_SPECIAL_MOTAE],
  [normalizeKey('motae'), LP_TYPE_SPECIAL_MOTAE],
  [normalizeKey(LP_TYPE_SPECIAL_NONG_MOTAE), LP_TYPE_SPECIAL_NONG_MOTAE],
  [normalizeKey('농모태'), LP_TYPE_SPECIAL_NONG_MOTAE],
  [normalizeKey('nong-motae'), LP_TYPE_SPECIAL_NONG_MOTAE],
  [normalizeKey('nongmotae'), LP_TYPE_SPECIAL_NONG_MOTAE],
  [normalizeKey(LP_TYPE_SPECIAL_GROWTH), LP_TYPE_SPECIAL_GROWTH],
  [normalizeKey('성장금융'), LP_TYPE_SPECIAL_GROWTH],
  [normalizeKey('growth-finance'), LP_TYPE_SPECIAL_GROWTH],
  [normalizeKey('growthfinance'), LP_TYPE_SPECIAL_GROWTH],
  [normalizeKey(LP_TYPE_SPECIAL_OTHER_POLICY), LP_TYPE_SPECIAL_OTHER_POLICY],
  [normalizeKey('특별조합원'), LP_TYPE_SPECIAL_OTHER_POLICY],
  [normalizeKey('정책출자자'), LP_TYPE_SPECIAL_OTHER_POLICY],
  [normalizeKey('정책자금'), LP_TYPE_SPECIAL_OTHER_POLICY],
  [normalizeKey('기타정책자금'), LP_TYPE_SPECIAL_OTHER_POLICY],
  [normalizeKey('정부기관'), LP_TYPE_SPECIAL_OTHER_POLICY],
  [normalizeKey('government'), LP_TYPE_SPECIAL_OTHER_POLICY],
])

export function normalizeLpType(value: string | null | undefined): string | null {
  const trimmed = (value || '').trim()
  if (!trimmed) return null
  return LP_TYPE_ALIASES.get(normalizeKey(trimmed)) ?? trimmed
}

export function normalizeLpTypeOrFallback(
  value: string | null | undefined,
  fallback: string = DEFAULT_LP_TYPE,
): string {
  return normalizeLpType(value) ?? fallback
}

export function labelLpType(value: string | null | undefined): string {
  const normalized = normalizeLpType(value)
  return normalized || '-'
}

export function groupLpType(value: string | null | undefined): string {
  const normalized = normalizeLpType(value)
  if (!normalized) return '-'
  if (normalized === LP_TYPE_INSTITUTIONAL || normalized === LP_TYPE_INDIVIDUAL) return '유한책임조합원'
  if (normalized === LP_TYPE_GP) return '업무집행조합원'
  if (
    normalized === LP_TYPE_SPECIAL_MOTAE ||
    normalized === LP_TYPE_SPECIAL_NONG_MOTAE ||
    normalized === LP_TYPE_SPECIAL_GROWTH ||
    normalized === LP_TYPE_SPECIAL_OTHER_POLICY
  ) {
    return '특별조합원'
  }
  return normalized
}

export function isGpLpType(value: string | null | undefined): boolean {
  return normalizeLpType(value) === LP_TYPE_GP
}

export function isKnownLpType(value: string | null | undefined): boolean {
  const normalized = normalizeLpType(value)
  return normalized != null && LP_TYPE_SET.has(normalized)
}
